Update finch to 0.17, publish async APIs

dlindenkreuz opened the pull request:
[Finch 0.17.0 was just released](https://github.com/sneako/finch/commit/603058bf67b83b4974fb516387cacff47312dcce)! 

This unblocks `Req.async_request` and allows us to `{:halt, ...}` from the `:into` callback option of `Req.request()`.

This PR publishes the features made possible by Finch 0.17.0's `stream_while` and removes the related pre-0.17 version checks. I am not entirely familiar with the Finch codebase, so there might be some accidental leftovers even if tests pass.

dlindenkreuz commented:
Cross-linking the related upstream issue for reference: https://github.com/sneako/finch/issues/237

wojtekmach commented on lib/req.ex:
```diff
@@ -962,14 +961,10 @@ defmodule Req do
 
-  # TODO: Wait for Finch 0.17
-  @doc false
   def async_request(request, options \\ []) do
```
Could you keep all these new function doc false for now? I need to do a pass on their docs anyway.

wojtekmach commented on lib/req/steps.ex:
```diff
@@ -882,28 +882,8 @@ defmodule Req.Steps do
-
-      Finch.stream(finch_req, finch_name, acc, fun, finch_options)
-    end
+  defp finch_stream_while(finch_req, finch_name, acc, fun, finch_options) do
```
Could you remove this defp and use Finch.stream_while directly at the callsite?

wojtekmach commented on mix.lock:
```diff
@@ -8,7 +8,7 @@
   "ex_doc": {:hex, :ex_doc, "0.31.0", "06eb1dfd787445d9cab9a45088405593dd3bb7fe99e097eaa71f37ba80c7a676", [:mix], [{:earmark_parser, "~> 1.4.39", [hex: :earmark_parser, repo: "hexpm", optional: false]}, {:makeup_c, ">= 0.1.1", [hex: :makeup_c, repo: "hexpm", optional: true]}, {:makeup_elixir, "~> 0.14", [hex: :makeup_elixir, repo: "hexpm", optional: false]}, {:makeup_erlang, "~> 0.1", [hex: :make ...
   "ezstd": {:hex, :ezstd, "1.0.8", "13584a08f2711b7b20173a41062f9c9454e255face6a30802fb676330db43495", [:rebar3], [], "hexpm", "a450e855d207cfee18263e78837964e59fe0cbed0dd99cb5e3755ebff7823171"},
-  "finch": {:hex, :finch, "0.16.0", "40733f02c89f94a112518071c0a91fe86069560f5dbdb39f9150042f44dcfb1a", [:mix], [{:castore, "~> 0.1 or ~> 1.0", [hex: :castore, repo: "hexpm", optional: false]}, {:mime, "~> 1.0 or ~> 2.0", [hex: :mime, repo: "hexpm", optional: false]}, {:mint, "~> 1.3", [hex: :mint, repo: "hexpm", optional: false]}, {:nimble_options, "~> 0.4 or ~> 1.0", [hex: :nimble_options, repo ...
+  "finch": {:hex, :finch, "0.17.0", "17d06e1d44d891d20dbd437335eebe844e2426a0cd7e3a3e220b461127c73f70", [:mix], [{:castore, "~> 0.1 or ~> 1.0", [hex: :castore, repo: "hexpm", optional: false]}, {:mime, "~> 1.0 or ~> 2.0", [hex: :mime, repo: "hexpm", optional: false]}, {:mint, "~> 1.3", [hex: :mint, repo: "hexpm", optional: false]}, {:nimble_options, "~> 0.4 or ~> 1.0", [hex: :nimble_options, repo ...
```
Could you change Finch version requirement in mix.exs too?

wojtekmach commented:
There is one other Finch change, deprecating :protocol in favor of :protocols. Finch defaults to just :http1 but Req can default to [:http1, :http2]. Lmk if you’d like to look into that but no pressure or rush of course!

dlindenkreuz commented:
Took a shot at deprecating `:protocol` in favor of `:protocols`. I added a runtime deprecation warning and renamed the field in the docs as well, but this likely needs to be pointed out more explicitly.

dlindenkreuz commented:
Feel free to adjust to your liking!

dlindenkreuz commented:
Just saw that `:into` docs should be updated as well to mention that `{:halt, …}` can be returned. Will do when I get back to my desk.

wojtekmach commented:
Thank you!

rubas commented:
@wojtekmach 

I would appreciate it, if you could create a new release with this fix as we are getting spammed with warnings.


> There is one other Finch change, deprecating :protocol in favor of :protocols. Finch defaults to just :http1 but Req can default to [:http1, :http2]. Lmk if you’d like to look into that but no pressure or rush of course!

```
warning: :protocol option is deprecated. Use :protocols instead.
  (nimble_options 1.1.0) lib/nimble_options.ex:572: NimbleOptions.validate_value/3
  (nimble_options 1.1.0) lib/nimble_options.ex:558: NimbleOptions.validate_option/3
  (nimble_options 1.1.0) lib/nimble_options.ex:540: NimbleOptions.reduce_options/2
  (elixir 1.16.1) lib/enum.ex:4842: Enumerable.List.reduce/3
  (elixir 1.16.1) lib/enum.ex:2582: Enum.reduce_while/3
  (nimble_options 1.1.0) lib/nimble_options.ex:533: NimbleOptions.validate_options/2
  (nimble_options 1.1.0) lib/nimble_options.ex:508: NimbleOptions.validate_options_with_schema_and_path/3
  (finch 0.18.0) lib/finch.ex:232: Finch.cast_pool_opts/1
  (finch 0.18.0) lib/finch.ex:201: anonymous fn/2 in Finch.pool_options!/1
  (stdlib 5.2) maps.erl:416: :maps.fold_1/4
  (finch 0.18.0) lib/finch.ex:160: Finch.start_link/1
```

dlindenkreuz commented:
Looks like this was released [in 0.4.9](https://github.com/wojtekmach/req/blob/main/CHANGELOG.md#v049-2024-02-14)!
